src/util/address.rs

```diff
@@ -511,13 +512,59 @@ impl Address {
         }
     }
 
-    /// Create a pay to taproot address
-    pub fn p2tr(taptweaked_key: schnorrsig::PublicKey, network: Network) -> Address {
+
+    /// Create a pay to taproot address for a given internal key and merkle root of tascript,
+    /// performing required tweaks according to BIP-342.
+    ///
+    /// Taproot (BIP-342) requires the keys to be tweaked with either tapscript merkle root or
+    /// self-tweaked with its own tagged hash. This function constructs non-tapscript based P2TR
+    /// address from an original untweaked public key value ("internal key"), performing the
+    /// necessary tweak with the merkle root of a tapscript.
+    ///
+    /// For generating P2TR addresses based on single public key please use
+    /// [`Address::p2tr_internal_key`] and [`Address::p2tr_tweaked_key`] methods.
+    #[inline]
+    pub fn p2tr_script<C: Verification>(secp: &Secp256k1<C>, mut internal_key: schnorrsig::PublicKey, merkle_root: TapBranchHash, network: Network) -> Address {
+        internal_key.script_tweak(&secp, merkle_root);
```

Comment:
This API is incorrect and can panic on user-crafted inputs. This also does not follow the BIP341 correctly. This should be H_taptweak(P||Hbranch(S)) . 

I have implemented similar things in #677 . I think this PR can wait on #677 where you can directly use some of the functions there.